### ISSUE TITLE
dataclasses: optional generic fields with non-None defaults

### ISSUE THREAD

**Issue by eltoder** [python/mypy#15887]
**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.11&gist=bcf4de945268517a0cd8b6b9835e852f
```python
from dataclasses import dataclass, field
from typing import Generic, TypeVar

T = TypeVar("T")

class V(Generic[T]):
    def __init__(self, *v: T) -> None:
        pass

class C:
    x: dict[str, int] | None = field(default_factory=dict)
    y: V[int] | None = field(default=V())
```

**Expected Behavior**

This should produce no errors.

**Actual Behavior**

```
main.py:11: error: Incompatible types in assignment (expression has type "dict[_KT, _VT]", variable has type "dict[str, int] | None")  [assignment]
main.py:12: error: Incompatible types in assignment (expression has type "V[<nothing>]", variable has type "V[int] | None")  [assignment]
Found 2 errors in 1 file (checked 1 source file)
```

If I remove `| None`, mypy works as expected. However, AFAICT, outside of dataclasses such assignments are accepted.

**Your Environment**

- Mypy version used: 1.5.0
- Python version used: 3.11

**Comment by eltoder** [comment]
Also, if I change `y` to use direct assignment instead of using `field`, the second error disappears:
```python
class C:
    x: dict[str, int] | None = field(default_factory=dict)
    y: V[int] | None = V()  # no error this way
```

**Comment by erictraut** [comment]
Here's a workaround that type checks without errors:

```python
class C:
    x: dict[str, int] | None = field(default_factory=dict[str, int])
    y: V[int] | None = field(default=V[int]())
```

**Comment by Viicos** [comment]
Can be closed as it seems to work as expected on mypy 1.13.

**Comment by brianschubert** [comment]
Thanks! FTR, this was fixed #16345
